mentat/agent_handler.py

```diff
@@ -0,0 +1,160 @@
+        self.agent_command_prompt = read_prompt(agent_command_prompt_path)
+
+    # Make this property readonly because we have to set things when we enable agent mode
+    @property
+    def agent_enabled(self):
+        return self._agent_enabled
+
+    def disable_agent_mode(self):
+        self._agent_enabled = False
+
+    async def enable_agent_mode(self):
+        ctx = SESSION_CONTEXT.get()
+
+        self._agent_enabled = True
+        ctx.stream.send(
+            "Finding files to determine how to test changes...", color="cyan"
+        )
+        features = ctx.code_context.get_all_features(CodeMessageLevel.FILE_NAME)
+        messages: List[ChatCompletionMessageParam] = [
+            ChatCompletionSystemMessageParam(
+                role="system", content=self.agent_file_selection_prompt
+            ),
+            ChatCompletionSystemMessageParam(
+                role="system",
+                content="\n".join(
+                    str(feature.path.relative_to(ctx.git_root)) for feature in features
+                ),
+            ),
+        ]
+        model = ctx.config.model
+        response = await ctx.llm_api_handler.call_llm_api(messages, model, False)
+        content = response.choices[0].message.content or ""
+
+        paths = [
+            Path(path) for path in content.strip().split("\n") if Path(path).exists()
+        ]
+        self.agent_file_message = ""
+        for path in paths:
+            file_contents = "\n".join(ctx.code_file_manager.read_file(path))
+            self.agent_file_message += f"{path}\n\n\n{file_contents}"
+
+        ctx.stream.send(
+            "The model has chosen these files to help it determine how to test its"
+            " changes:",
+            color="cyan",
+        )
+        ctx.stream.send("\n".join(str(path) for path in paths))
+        ctx.cost_tracker.log_api_call_stats(
+            prompt_tokens(messages, model),
+            count_tokens(content, model, full_message=False),
+            model,
+        )
+
+        messages.append(
+            ChatCompletionAssistantMessageParam(role="assistant", content=content)
+        )
+        ctx.conversation.add_transcript_message(
+            ModelMessage(message=content, prior_messages=messages, message_type="agent")
+        )
+
+    async def _determine_commands(self):
+        ctx = SESSION_CONTEXT.get()
+
+        model = ctx.config.model
+        code_message = await ctx.code_context.get_code_message()
+        messages = ctx.conversation.get_messages() + [
+            ChatCompletionSystemMessageParam(
+                role="system", content=self.agent_command_prompt
+            ),
+            ChatCompletionSystemMessageParam(role="system", content=code_message),
+            ChatCompletionSystemMessageParam(
+                role="system", content=self.agent_file_message
+            ),
+        ]
+        # TODO: Should this even be a separate call or should we collect commands in the edit call?
```

Comment:
I don't think so because we need a preset list of functions we want to call; the whole point of this call is that we don't know what commands we can run and the llm is figuring it out.